tests/components/plugwise/conftest.py

```diff
@@ -0,0 +1,167 @@
+"""Setup mocks for the Plugwise integration tests."""
+
+import os
+import pickle
+import re
+
+from Plugwise_Smile.Smile import Smile
+import pytest
+
+from tests.async_mock import AsyncMock, patch
+from tests.test_util.aiohttp import AiohttpClientMocker
+
+global ENVIRONMENT
+
+
+def _read_pickle(environment, call):
+    """Uncompress the pickle data."""
+    fixture = call + ".pickle"
+    path = os.path.join(
+        os.path.dirname(__file__), "../../fixtures/plugwise/" + environment, fixture
+    )
+    with open(path, "rb") as fixture_file:
+        return pickle.load(fixture_file)
+
+
+@pytest.fixture(name="mock_smile")
+def mock_smile():
+    """Create a Mock Smile for testing exceptions."""
+    with patch(
+        "homeassistant.components.plugwise.config_flow.Smile",
+    ) as smile_mock:
+        smile_mock.InvalidAuthentication = Smile.InvalidAuthentication
+        smile_mock.ConnectionFailedError = Smile.ConnectionFailedError
+        smile_mock.return_value.connect.return_value = True
+        yield smile_mock.return_value
+
+
+@pytest.fixture(name="mock_smile_unauth")
+def mock_smile_unauth(aioclient_mock: AiohttpClientMocker) -> None:
+    """Mock the Plugwise Smile unauthorized for Home Assistant."""
+    aioclient_mock.get(re.compile(".*"), status=401)
+    aioclient_mock.put(re.compile(".*"), status=401)
+
+
+@pytest.fixture(name="mock_smile_error")
+def mock_smile_error(aioclient_mock: AiohttpClientMocker) -> None:
+    """Mock the Plugwise Smile server failure for Home Assistant."""
+    aioclient_mock.get(re.compile(".*"), status=500)
+    aioclient_mock.put(re.compile(".*"), status=500)
+
+
+@pytest.fixture(name="mock_smile_notconnect")
+def mock_smile_notconnect():
+    """Mock the Plugwise Smile general connection failure for Home Assistant."""
+    with patch("homeassistant.components.plugwise.Smile") as smile_mock:
+        smile_mock.InvalidAuthentication = Smile.InvalidAuthentication
+        smile_mock.ConnectionFailedError = Smile.ConnectionFailedError
+        smile_mock.PlugwiseError = Smile.PlugwiseError
+        smile_mock.return_value.connect.side_effect = AsyncMock(return_value=False)
+        yield smile_mock.return_value
+
+
+def _get_device_data(device_id):
+    """Mock return data for specific devices."""
+    global ENVIRONMENT
```

Comment:
Please don't use global. There should be other ways of solving this.

1. Try to use `functools.partial` to set the environment as an argument to `_get_device_data`.
2. Make a factory function that takes the environment as a parameter and return `_get_device_data` from the factory function.

Reply:
Duly noted, I wasn't pleased with it but learned how to properly use partial, tnx